### PR TITLE
Support for Binary Classification with PGD

### LINKED ISSUE #306: Binary Classification with Single Output Neuron
Hello,

1. Does ART support generating adversarial examples on neural networks designed for binary classification with a single output neuron?
2. Are there any examples in PyTorch for how to use a pretrained classifier? All examples I've seen include training using the PyTorchClassifier wrapper. I'm not sure how to instantiate one of these objects if I've already got a trained model.

Thanks!

### PR DESCRIPTION
**Pull request by TrojAISec**
# Description

Please include a summary of the change, motivation and which issue is fixed. Any dependencies changes should also be included.

Fixes #306

## Type of change

Please check all relevant options.

- [X] Improvement (non-breaking)
- [ ] Bug fix (non-breaking)
- [X] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Testing

[X] test_pytorch_binary

Please describe the tests that you ran to verify your changes. Consider listing any relevant details of your test configuration.

- [ ] Test A
- [ ] Test B

**Test Configuration**:
- OS
- Python version: 3.8.5
- ART version or commit number: dev_1.7.0
- PyTorch version: 1.8.1+cpu

# Checklist

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code
- [ ] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes

This is my first commit, if anything needs changing, please let me know! If it looks okay, we can move forward with adding the support for more frameworks than just PyTorch.

### CODE REVIEW

File: art/estimators/classification/pytorch.py [hunk 1]
```diff
@@ -717,6 +723,9 @@ def loss_gradient(  # pylint: disable=W0221
         else:
             labels_t = y_preprocessed
 
+        # if labels_t is None:
+        #     labels_t = y
+
```
Comment (beat-buesser) [inline art/estimators/classification/pytorch.py:728]: Are these lines still required?

File: art/utils.py [hunk 2]
```diff
@@ -620,6 +625,7 @@ def compute_success_array(
 ) -> float:
     """
     Compute the success rate of an attack based on clean samples, adversarial samples and targets or correct labels.
+    Added shape check for binary case
```
Comment (beat-buesser) [inline art/utils.py:628]:
```suggestion
```

File: tests/classifiersFrameworks/test_tensorflow.py [hunk 3]
```diff
@@ -23,10 +23,15 @@
...
-from art.attacks.evasion import FastGradientMethod
+from art.attacks.evasion import FastGradientMethod, ProjectedGradientDescent
 
 from tests.attacks.utils import backend_test_defended_images
 from tests.utils import ARTTestException
+import art.estimators.classification
+import art
+import sklearn
+import sklearn.datasets
+import sklearn.model_selection
```
Comment (beat-buesser) [inline tests/classifiersFrameworks/test_tensorflow.py:34]: To follow PEP8, could you please move the imports from `sklearn` to line 22 and the import from `art` to line 27?

File: art/estimators/classification/pytorch.py [hunk 4]
```diff
@@ -943,11 +954,13 @@ def forward(self, x):
                         result = []
                         if isinstance(self._model, nn.Sequential):
                             for _, module_ in self._model._modules.items():
-                                x = module_(x)
+                                x = module_(x.float())
```
Comment (beat-buesser) [inline art/estimators/classification/pytorch.py:957]: Why did you add this `float()`?

File: art/estimators/classification/pytorch.py [hunk 5]
```diff
@@ -943,11 +954,13 @@ def forward(self, x):
...
                                 result.append(x)
 
                         elif isinstance(self._model, nn.Module):
-                            x = self._model(x)
+                            x = self._model(x.float())
+                            if len(x.shape) > 1 and self.nb_classes == 2:
+                                x = x.squeeze(dim=1)
```
Comment (beat-buesser) [inline art/estimators/classification/pytorch.py:963]: Why is this step necessary?

File: art/utils.py [hunk 6]
```diff
@@ -630,12 +636,18 @@ def compute_success_array(
...
+        preds = np.argmax(preds, axis=1)
+    else:
+        adv_preds = np.round(adv_preds)
+        preds = np.round(preds)
     if targeted:
         attack_success = adv_preds == np.argmax(labels, axis=1)
-    else:
-        preds = np.argmax(classifier.predict(x_clean, batch_size=batch_size), axis=1)
-        attack_success = adv_preds != preds
+    # preds = np.argmax(classifier.predict(x_clean, batch_size=batch_size), axis=1)
```
Comment (beat-buesser) [inline art/utils.py:649]: I think we can now also remove this commented line from the existing function.
```suggestion
```

File: art/utils.py [hunk 7]
```diff
@@ -630,12 +636,18 @@ def compute_success_array(
...
+    adv_preds = classifier.predict(x_adv, batch_size=batch_size)
+    if len(preds.shape) >= 2:
+        adv_preds = np.argmax(adv_preds, axis=1)
+        preds = np.argmax(preds, axis=1)
+    else:
+        adv_preds = np.round(adv_preds)
+        preds = np.round(preds)
     if targeted:
         attack_success = adv_preds == np.argmax(labels, axis=1)
-    else:
```
Comment (beat-buesser) [inline art/utils.py:636]: We need to make sure that `attack_success` for `targeted` does not get overwritten later in line 650.

File: art/utils.py [hunk 8]
```diff
@@ -630,12 +636,18 @@ def compute_success_array(
     :param batch_size: Batch size.
     :return: Percentage of successful adversarial samples.
     """
-    adv_preds = np.argmax(classifier.predict(x_adv, batch_size=batch_size), axis=1)
+    preds = classifier.predict(x_clean, batch_size=batch_size)
```
Comment (beat-buesser) [inline art/utils.py:639]: We should be able to save the computation to evaluate `x_clean` for `targeted` and only evaluate `x_adv`.

File: tests/classifiersFrameworks/test_pytorch.py [hunk 9]
```diff
@@ -221,3 +226,91 @@ def test_fgsm_defences(art_warning, fix_get_mnist_subset, image_dl_estimator, de
...
+        attacker = art.attacks.evasion.ProjectedGradientDescent(classifier, eps=0.5)
+        generated = attacker.generate(test_x_batch)
+        adv_predicted = classifier.predict(generated)
+        assert (adv_predicted != preds).all()
+    except ARTTestException as e:
+        art_warning(e)
+
+
+# # @pytest.mark.only_with_platform("pytorch")
+# def test_pytorch_binary_FGSM(art_warning):
```
Comment (beat-buesser) [inline tests/classifiersFrameworks/test_pytorch.py:278]: Can we uncomment this test?

File: art/estimators/classification/pytorch.py (hunk 1 above)
Comment (TrojAISec) [reply art/estimators/classification/pytorch.py]: Removed.

File: art/utils.py (hunk 2 above)
Comment (TrojAISec) [reply art/utils.py]: Removed.

File: tests/classifiersFrameworks/test_pytorch.py [hunk 10]
```diff
@@ -28,6 +26,13 @@
 from tests.attacks.utils import backend_test_defended_images
 from tests.utils import ARTTestException
 
+import torch
+import torch.nn.functional as F
+import torch.nn as nn
+import torch.optim as optim
+import art.estimators.classification
+import sklearn.datasets
```
[the remark this answers is not in the document]
Comment (TrojAISec) [reply tests/classifiersFrameworks/test_pytorch.py]: Completed, hopefully is as expected. If not let us know.

File: tests/classifiersFrameworks/test_tensorflow.py (hunk 3 above)
Comment (TrojAISec) [reply tests/classifiersFrameworks/test_tensorflow.py]: Completed, hopefully is as expected. If not let us know.

File: art/utils.py (hunk 6 above)
Comment (TrojAISec) [reply art/utils.py]: Removed.

File: art/utils.py (hunk 7 above)
Comment (TrojAISec) [reply art/utils.py:636]: Tentatively added back else to avoid overwrite of var, waiting for pytest results...

```
    if targeted:
        attack_success = adv_preds == np.argmax(labels, axis=1)
    else:
        attack_success = adv_preds != preds
```

File: art/utils.py (hunk 8 above)
Comment (TrojAISec) [reply art/utils.py]: Committed change to compute only adv_preds first and return attack_success before moving on to clean preds. Not sure if most elegant solution, but should work. Waiting on pytest.

File: tests/classifiersFrameworks/test_pytorch.py (hunk 9 above)
Comment (TrojAISec) [reply tests/classifiersFrameworks/test_pytorch.py:279]: Removed, FGSM proved difficult due to some dtype conflicts in the loss + generate function. Will try and get a more thorough description of the issue from Max at some point.

File: tests/classifiersFrameworks/test_pytorch.py [hunk 11]
```diff
@@ -28,6 +32,8 @@
 from tests.attacks.utils import backend_test_defended_images
 from tests.utils import ARTTestException
 
+import sklearn.datasets
+
```
Comment (beat-buesser) [inline tests/classifiersFrameworks/test_pytorch.py:36]:
```suggestion
```

File: tests/classifiersFrameworks/test_pytorch.py [hunk 12]
```diff
@@ -221,3 +227,91 @@ def test_fgsm_defences(art_warning, fix_get_mnist_subset, image_dl_estimator, de
...
+            x = torch.sigmoid(self.layer_2(x))
+
+            return x
+
+    try:
+        device = "cpu"
+        x, y = sklearn.datasets.make_classification(
+            n_samples=10000, n_features=20, n_informative=5, n_redundant=2, n_repeated=0, n_classes=2
+        )
+        train_x, test_x, train_y, test_y = sklearn.model_selection.train_test_split(x, y, test_size=0.2)
```
Comment (beat-buesser) [inline tests/classifiersFrameworks/test_pytorch.py:256]:
```suggestion
        (x_train_mnist, y_train_mnist), (x_test_mnist, y_test_mnist) = get_mnist_dataset
```

File: tests/classifiersFrameworks/test_pytorch.py [hunk 13]
```diff
@@ -221,3 +227,91 @@ def test_fgsm_defences(art_warning, fix_get_mnist_subset, image_dl_estimator, de
         backend_test_defended_images(attack, fix_get_mnist_subset)
     except ARTTestException as e:
         art_warning(e)
+
+
+@pytest.mark.only_with_platform("pytorch")
+def test_pytorch_binary_PGD(art_warning):
```
Comment (beat-buesser) [inline tests/classifiersFrameworks/test_pytorch.py:233]:
```suggestion
def test_pytorch_binary_PGD(art_warning, get_mnist_dataset):
```

File: tests/classifiersFrameworks/test_pytorch.py (hunk 11 above)
Comment (beat-buesser) [inline tests/classifiersFrameworks/test_pytorch.py:36]: Could you please move this line to after line 24 for PEP8?

File: art/estimators/classification/pytorch.py [hunk 14]
```diff
@@ -311,7 +317,7 @@ def predict(  # pylint: disable=W0221
             )
 
             with torch.no_grad():
-                model_outputs = self._model(torch.from_numpy(x_preprocessed[begin:end]).to(self._device))
+                model_outputs = self._model(torch.from_numpy(x_preprocessed[begin:end]).float().to(self._device))
```
Comment (beat-buesser) [inline art/estimators/classification/pytorch.py:320]:
```suggestion
                model_outputs = self._model(torch.from_numpy(x_preprocessed[begin:end]).to(self._device))
```

File: tests/classifiersFrameworks/test_pytorch.py (hunk 12 above)
Comment (beat-buesser) [inline tests/classifiersFrameworks/test_pytorch.py:256]:
```suggestion
        _, test_x, _, test_y = sklearn.model_selection.train_test_split(x, y, test_size=0.2)
        test_x = test_x.astype(np.float32)
```

File: tests/classifiersFrameworks/test_pytorch.py [hunk 15]
```diff
@@ -221,3 +227,91 @@ def test_fgsm_defences(art_warning, fix_get_mnist_subset, image_dl_estimator, de
...
+#         def forward(self, x):
+#             x = F.relu(self.layer_1(x))
+#             x= torch.sigmoid(self.layer_2(x))
+
+#             return x
+#     try:
+#         device = "cpu"
+#         x, y = sklearn.datasets.make_classification(n_samples=10000, n_features=20, n_informative=5, n_redundant=2,
+#                                                     n_repeated=0, n_classes=2)
+#         train_x, test_x, train_y, test_y = sklearn.model_selection.train_test_split(x,y, test_size=0.2)
```
Comment (beat-buesser) [inline tests/classifiersFrameworks/test_pytorch.py:298]:
```suggestion
#        _, test_x, _, test_y = sklearn.model_selection.train_test_split(x, y, test_size=0.2)
#        test_x = test_x.astype(np.float32)
```

File: tests/classifiersFrameworks/test_tensorflow.py [hunk 16]
```diff
@@ -215,3 +221,57 @@ def test_fgsm_defences(art_warning, fix_get_mnist_subset, image_dl_estimator):
...
+
+
+@pytest.mark.only_with_platform("tensorflow2")
+def test_binary_keras_instantiation_and_attack_PGD(art_warning):
+    tf.compat.v1.disable_eager_execution()
+    try:
+        x, y = sklearn.datasets.make_classification(
+            n_samples=10000, n_features=20, n_informative=5, n_redundant=2, n_repeated=0, n_classes=2
+        )
+        train_x, test_x, train_y, test_y = sklearn.model_selection.train_test_split(x, y, test_size=0.2)
```
Comment (beat-buesser) [inline tests/classifiersFrameworks/test_tensorflow.py:233]:
```suggestion
        _, test_x, _, test_y = sklearn.model_selection.train_test_split(x, y, test_size=0.2)
        test_x = test_x.astype(np.float32)
```

File: tests/classifiersFrameworks/test_tensorflow.py [hunk 17]
```diff
@@ -215,3 +221,57 @@ def test_fgsm_defences(art_warning, fix_get_mnist_subset, image_dl_estimator):
...
+
+
+@pytest.mark.only_with_platform("tensorflow2")
+def test_binary_tf2_instantiation_and_attack_PGD(art_warning):
+    tf.compat.v1.disable_eager_execution()
+    try:
+        x, y = sklearn.datasets.make_classification(
+            n_samples=10000, n_features=20, n_informative=5, n_redundant=2, n_repeated=0, n_classes=2
+        )
+        train_x, test_x, train_y, test_y = sklearn.model_selection.train_test_split(x, y, test_size=0.2)
```
Comment (beat-buesser) [inline tests/classifiersFrameworks/test_tensorflow.py:260]:
```suggestion
        train_x, test_x, train_y, test_y = sklearn.model_selection.train_test_split(x, y, test_size=0.2)
        train_x = train_x.astype(np.float32)
        test_x = test_x.astype(np.float32)
```

File: tests/classifiersFrameworks/test_pytorch.py [hunk 18]
```diff
@@ -221,3 +227,91 @@ def test_fgsm_defences(art_warning, fix_get_mnist_subset, image_dl_estimator, de
...
+        opt = optim.Adam(model.parameters(), lr=0.001)
+        classifier = art.estimators.classification.PyTorchClassifier(
+            model=model,
+            loss=loss_func,
+            optimizer=opt,
+            input_shape=(1, 28, 28),
+            nb_classes=2,
+        )
+        test_x_batch = test_x[0:16]
+        preds = classifier.predict(test_x_batch)
```
Comment (beat-buesser) [inline tests/classifiersFrameworks/test_pytorch.py:269]: Should we fit the classifier?

File: art/estimators/classification/pytorch.py [hunk 19]
```diff
@@ -269,17 +269,20 @@ def reduce_labels(self, y: Union[np.ndarray, "torch.Tensor"]) -> Union[np.ndarra
...
-            return np.argmax(y, axis=1)
-
-        if self._reduce_labels:  # float labels
-            if isinstance(y, torch.Tensor):
-                return torch.argmax(y, dim=1).type("torch.FloatTensor")
-            y_index = np.argmax(y, axis=1).astype(np.float32)
-            y_index = np.expand_dims(y_index, axis=1)
-            return y_index
+        # Checking for exactly 2 classes to support binary classification
+        if self.nb_classes > 2 or self.nb_classes == 1:
```
Comment (beat-buesser) [inline art/estimators/classification/pytorch.py:273]:
```suggestion
        if self.nb_classes > 2:
```

File: art/utils.py [hunk 20]
```diff
@@ -513,7 +513,9 @@ def check_and_transform_label_format(
     :return: Labels with shape `(nb_samples, nb_classes)` (one-hot) or `(nb_samples,)` (index).
     """
     if labels is not None:
-        if len(labels.shape) == 2 and labels.shape[1] > 1:
+        if nb_classes == 1:
+            pass
```
Comment (beat-buesser) [inline art/utils.py:517]:
```suggestion
```

File: art/estimators/classification/pytorch.py [hunk 21]
```diff
@@ -388,6 +400,8 @@ def fit(self, x: np.ndarray, y: np.ndarray, batch_size: int = 128, nb_epochs: in
                 model_outputs = self._model(i_batch)
 
                 # Form the loss function
+                print(model_outputs[-1].dtype)
+                print(o_batch.dtype)
```
Comment (beat-buesser) [inline art/estimators/classification/pytorch.py:404]:
```suggestion
```

File: art/attacks/evasion/fast_gradient.py [hunk 22]
```diff
@@ -231,7 +231,13 @@ def generate(self, x: np.ndarray, y: Optional[np.ndarray] = None, **kwargs) -> n
                 # Use model predictions as correct outputs
                 logger.info("Using model predictions as correct labels for FGM.")
                 y = get_labels_np_array(self.estimator.predict(x, batch_size=self.batch_size))  # type: ignore
-            y = y / np.sum(y, axis=1, keepdims=True)
+
+            if self.estimator.nb_classes == 2 and y.shape[1] == 1:
+                axis = 0
```
Comment (beat-buesser) [inline art/attacks/evasion/fast_gradient.py:236]: I'm not sure if we need to do anything for this case. Would `axis=0` sum over all samples in the binary case?

File: art/attacks/evasion/fast_gradient.py [hunk 23]
```diff
@@ -231,7 +231,13 @@ def generate(self, x: np.ndarray, y: Optional[np.ndarray] = None, **kwargs) -> n
...
+            else:
+                axis = 1
+
+            y = y / np.sum(y, axis=axis, keepdims=True)
```
Comment (beat-buesser) [inline art/attacks/evasion/fast_gradient.py:240]:
```suggestion
            if self.estimator.nb_classes > 2:
                y = y / np.sum(y, axis=1, keepdims=True)
```

File: art/estimators/classification/pytorch.py [hunk 24]
```diff
@@ -315,8 +321,14 @@ def predict(  # pylint: disable=W0221
             with torch.no_grad():
                 model_outputs = self._model(torch.from_numpy(x_preprocessed[begin:end]).to(self._device))
             output = model_outputs[-1]
-            results[begin:end] = output.detach().cpu().numpy()
+            # results[begin:end] = output.detach().cpu().numpy() #old
+            if len(output.shape) == 1:
+                output = np.expand_dims(output.detach().cpu().numpy(), axis=1).astype(np.float32)
+            else:
+                output = output.detach().cpu().numpy().astype(np.float32)
```
Comment (beat-buesser) [inline art/estimators/classification/pytorch.py:328]:
```suggestion
            output = output.detach().cpu().numpy().astype(np.float32)
            if len(output.shape) == 1:
                output = np.expand_dims(output).astype(np.float32)
```

File: art/estimators/classification/pytorch.py [hunk 25]
```diff
@@ -948,14 +963,15 @@ class ModelWrapper(nn.Module):
...
-                    def __init__(self, model: torch.nn.Module):
+                    def __init__(self, model: torch.nn.Module, nb_classes=nb_classes):
                         """
                         Initialization by storing the input model.
 
                         :param model: PyTorch model. The forward function of the model must return the logit output.
                         """
                         super().__init__()
                         self._model = model
+                        self.nb_classes = nb_classes
```
Comment (beat-buesser) [inline art/estimators/classification/pytorch.py:974]:
```suggestion
```

File: art/estimators/classification/pytorch.py [hunk 26]
```diff
@@ -935,6 +949,7 @@ def __repr__(self):
 
     def _make_model_wrapper(self, model: "torch.nn.Module") -> "torch.nn.Module":
         # Try to import PyTorch and create an internal class that acts like a model wrapper extending torch.nn.Module
+        nb_classes = self.nb_classes
```
Comment (beat-buesser) [inline art/estimators/classification/pytorch.py:952]:
```suggestion
```

File: art/utils.py [hunk 27]
```diff
@@ -630,11 +638,17 @@ def compute_success_array(
     :param batch_size: Batch size.
     :return: Percentage of successful adversarial samples.
     """
-    adv_preds = np.argmax(classifier.predict(x_adv, batch_size=batch_size), axis=1)
+    preds = classifier.predict(x_clean, batch_size=batch_size)
```
Comment (beat-buesser) [inline art/utils.py:641]: Can we move all lines related to `preds` into the `else` block after line 651 to save model evaluations?

**Review by beat-buesser — APPROVED** [review]
@TrojAISec @stroypet Thank you very much for your first of hopefully many contributions to ART providing support for binary classification with a single output in PyTorch, TensorFlow v2 and Keras classifiers and testing it with ProjectedGradientDescent!
